packages/vm/src/imports.rs

```diff
@@ -74,7 +74,8 @@ pub fn do_remove<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResul
     }
 
     let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
-    let used_gas = with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.remove(&key)?))?;
+    let (_, used_gas) =
```

Comment:
However, I think there is a big difference between these two (which we should make explicit).

For storage operations, they charge the sdk gas meter already. We track this by reducing the remaining limit, but *do not increase* the amount of gas consumed by wasm when reporting it.

For the API calls, we *do not charge sdk gas meter* but rather wish to act as if the wasm code used X points.

The fact both use `try_consume_gas` is confusing. I assume one should be `consume_wasm_gas` and the other `reduce_remaining_gas_limit` or such.

I don't have all the semantics in my head, but this is a key distinction we should make clear with naming. Maybe @reuvenpo can chime in on this point

Reply:
Hah! This means our gas used implementation in `do_read` and friends is wrong at the moment, as the gas is charged twice.

We need to sync back the gas consumption of the SDK calls in order to adjust the limit. Otherwise you can use the gas limit twice (in the SDK and in the VM). But in the VM we only have one gas meter.

> For the API calls, we _do not charge sdk gas meter_ but rather wish to act as if the wasm code used X points.

I think this behaviour is what we need for the DB calls as well. Everything coming from the VM needs to be tracked separately and goes on the `Instance`'s bill.

Reply:
Actually [`try_consume_gas`](https://github.com/CosmWasm/cosmwasm/blob/0fe88961f746bf42cb59878ea3db9895eafe141b/packages/vm/src/context.rs#L199-L225) is proper for the places where the cost is already charged in the cosmos-sdk gas meter.

It calls into `GasState::use_gas`: 

```rust
    fn use_gas(&mut self, amount: u64) {
        self.externally_used_gas += amount;
    }
```

and all calculations do a balancing app of limit, external gas, and wasmer used gas. I remember reviewing this code when @reuvenpo wrote it and it needed testing to become correct, but I am pretty sure this is all good.

What we need is a new call to charge gas for the API functions. I am talking about the code in `imports.rs`: [`do_canonicalize_address`](https://github.com/CosmWasm/cosmwasm/blob/34eb0e8df8dc05831a52fec9d007a2ca608af676/packages/vm/src/imports.rs#L102) and [`do_humanize_address`](https://github.com/CosmWasm/cosmwasm/blob/34eb0e8df8dc05831a52fec9d007a2ca608af676/packages/vm/src/imports.rs#L118).

They should not use the same `try_consume_gas` as the other ones.

Reply:
We currently do not expose such interface in the [`backend/singlepass`](https://github.com/CosmWasm/cosmwasm/blob/d04123d7496d7470546ed96d49cb9f45fea882e1/packages/vm/src/backends/singlepass.rs#L45-L63), only `set_gas_limit` and `get_gas_left`, but there is the proper function we could use: `metering::set_points_used`.

Maybe something like this:

```rust
fn charge_wasm_gas(instance: &mut Instance, gas: u64) {
    let used = metering::get_points_used(&instance);
    metering::set_points_used(instance, used+gas);
}
```

which should probably live inside the backends package

Reply:
Perfect, thanks for all the explanation. Makes sense. Naming needs some adjustmet to make this clearer.

Stepping back from the implementation and challenging the status quo: do we really want two different metering for storage and API calls into the Cosmos SDK? Can't we use the SDK metering for all of them or for none of them?

Reply:
If Cosmos SDK performs metering for storage operations, are those are incorrectly summed up because this line (in go-cosmwasm)

```rust
*gas_used = gas_limit - instance.get_gas_left();
```

includes both the gas used by the VM as well as the externally used gas?

Reply:
> If Cosmos SDK performs metering for storage operations, are those are incorrectly summed up because this line (in go-cosmwasm)

`*gas_used = gas_limit - instance.get_gas_left();`

> includes both the gas used by the VM as well as the externally used gas?

Probably true there is some double-charging. I do remember reporting a large jump in the sdk tests when updating to 0.9. Probably this was the reason. 

How do we do this properly? Do we need https://github.com/wasmerio/wasmer/pull/996 ?

Reply:
> How do we do this properly?

Add a method on Instance that returns the value we need directly: _internally used gas_, which is `(original_gas_limit - gas_left) - externally_used_gas`.